pkg/kudoctl/util/repo/package.go

```diff
@@ -0,0 +1,312 @@
+package repo
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/kudobuilder/kudo/pkg/apis/kudo/v1alpha1"
+	"github.com/kudobuilder/kudo/pkg/bundle"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	frameworkV0FileName = "-framework.yaml"
+	versionV0FileName   = "-frameworkversion.yaml"
+	instanceV0FileName  = "-instance.yaml"
+
+	frameworkV1FileName     = "framework.yaml"
+	templatesV1Folder       = "templates"
+	templateV1FileNameRegex = "templates/.*.yaml"
+	paramsV1FileName        = "params.yaml"
+)
+
+const apiVersion = "kudo.k8s.io/v1alpha1"
+
+// InstallCRDs is collection of CRDs that are used when installing framework
+// during installation, package format is converted to this structure
+type InstallCRDs struct {
+	Framework        *v1alpha1.Framework
+	FrameworkVersion *v1alpha1.FrameworkVersion
+	Instance         *v1alpha1.Instance
+}
+
+func isFrameworkV0File(name string) bool {
+	return strings.HasSuffix(name, frameworkV0FileName)
+}
+
+func isVersionV0File(name string) bool {
+	return strings.HasSuffix(name, versionV0FileName)
+}
+
+func isInstanceV0File(name string) bool {
+	return strings.HasSuffix(name, instanceV0FileName)
+}
+
+// ReadTarballPackage reads package from tarball and converts it to the CRD format
+func ReadTarballPackage(r io.Reader) (*InstallCRDs, error) {
+	p, err := untarV1Package(r)
+	if err != nil {
+		return nil, errors.Wrap(err, "while untarring package")
+	}
+	return p.getInstallCRDs()
+}
+
+// ReadFileSystemPackage reads package from filesystem and converts it to the CRD format
+func ReadFileSystemPackage(path string) (*InstallCRDs, error) {
+	p, err := fromFilesystem(path)
+	if err != nil {
+		return nil, errors.Wrap(err, "while reading package from filesystem")
+	}
+	return p.getInstallCRDs()
+}
+
+func fromFilesystem(packagePath string) (*v1Package, error) {
+	result := newV1Package()
+	err := filepath.Walk(packagePath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			// skip directories
+			return nil
+		}
+		relativePath := strings.TrimPrefix(path, packagePath)
+		if path == packagePath {
+			// skip the root folder, as Walk always starts there
+			return nil
+		}
+		bytes, err := ioutil.ReadFile(path)
+		if err != nil {
+			return err
+		}
+		switch {
```

Comment:
the code seems to conflate file walking and file processing.   It is reasonably easy to follow... but what do you think of having a filewalker that returns a set of string paths... and a processor that takes a set of string paths and processes them?     In my mind it simplifies the reasoning and testing of the 2 concerns.

Reply:
I see your point, but the filewalker is actually what the library provides. If you return string paths then you basically need to iterate that tarball two times. I saw a lot of code snippets how people deal with tarballs in golang and everybody seems to be doing it this way...